File: tests/co-simulations/proc/1/python/tb.py
import random
import traceback
import sys

import cosim
import afbd

WRITE_FIFO_PATH = sys.argv[1]
READ_FIFO_PATH = sys.argv[2]
REG_JSON = sys.argv[3]


try:
    iface = cosim.Iface(WRITE_FIFO_PATH, READ_FIFO_PATH)

    Main, _ = afbd.generate(iface, REG_JSON)

    a = random.randint(0, 2 ** 16 - 1)
    b = random.randint(0, 2 ** 16 - 1)

    print(f"Calling add function, a = {a}, b = {b}")
    Main.Add(a, b)

    print(f"Reading result")
    result = Main.Result.read()

    if a + b != result:
        print(f"Wrong result, got {result}, expecting {a+b}")
        iface.end(1)

    iface.end(0)

except Exception as E:
    iface.end(1)
    print(traceback.format_exc())

File: tests/co-simulations/paramProc/1/c-sync/tb.c
#include <assert.h>
#include <stdio.h>
#include <time.h>
#include <stdlib.h>

#include "cosim_iface.h"

#include "afbd.h"
#include "main.h"
#define AFBD_IFACE &iface


int main(int argc, char *argv[]) {
	assert(argc == 3);

	afbd_iface_t iface = cosim_iface_iface();

	cosim_iface_init(argv[1], argv[2], NULL);

	srand(time(NULL));
	const uint16_t a = rand() & 0xFFFF; // 16 bits value
	const uint16_t b = rand() & 0xFFFF; // 16 bits value

	printf("generated random values: a = %d, b = %d\n", a, b);

	printf("calling add function\n");

	uint32_t err;

	err = afbd_main_add(&iface, a, b);
	if (err) {
		fprintf(stderr, "error calling add function: %d", err);
		cosim_iface_end(1);
	}

	uint32_t result;
	err = afbd_read(main_result, &result);
	if (err) {
		fprintf(stderr, "error reading result: %d", err);
		cosim_iface_end(1);
	}

	if (a + b != result) {
		fprintf(stderr, "wrong result %d, expecting %d\n", result, a + b);
		cosim_iface_end(1);
	}

	cosim_iface_end(0);

	return 0;
}
